DAC Beeps

peabody124 opened the pull request:
This is code that will generate a series of audio beeps out of the DAC pin on
Sparky2 and BrushedSparky. This is pretty nice to pump into FPV audio so 
you can hear the RSSI and voltage (RSSI is encoded into repetition rate with
higher rate being low RSSI, the pitch of the beep drops as your voltage does).

The PIOS_DAC_COMMON file is to allow the FSK DAC code (under progress)
to coexist.

peabody124 committed:
Sparky2: enable DACBEEP mode

peabody124 committed:
PIOS DACBEEP: check in file

peabody124 committed:
PIOS DACBEEP: support changing output frequency

peabody124 committed:
Sparky2: move RFM22b SPI TX to DMA1_Stream7

This makes room for the DAC1 to use DMA_Stream5. This shouldn't
make any difference anyway since we aren't using the IRQ handler
for the SPI TX.

peabody124 committed:
PIOS DACBEEP: working now

peabody124 committed:
PIOS DACBEEP: ChBios IRQ prologue/epilogue

peabody124 committed:
PIOS DACBEEP: sound is 0.4V wave centered at 0.5V

This is the voltage range expected by VTXs and gets
rid of the harsh sound when it was not centered
properly.

peabody124 committed:
PIOS DACBEEP: use defined register address

peabody124 committed:
PIOS DACBEEP: bump copyright

peabody124 committed:
Sparky2: bump copyright

peabody124 committed:
Sparky2: allow selecting the DAC output function

There are three options:
 - ADC - functions as normal analog input
 - Beep - transmits beep codes as audio output
 - FSK Telemetry - Bel202 light weight telemetry

peabody124 committed:
PIOS DACBEEP: factor basic config and irq to common file

This is required for PIOS FSKDAC to share the DMA driver and
gets rid of some reundancy.

peabody124 committed:
PIOS DACBEEP: get rid of unused cfg structure

peabody124 committed:
PIOS DACBEEP: small cleanup

peabody124 committed:
Modules/DacBeepCodes: check in module

peabody124 commented:
See demonstration of video here https://www.instagram.com/p/-uypx1lqoH/

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+	// Start main task if it is enabled
+	if (module_enabled) {
+		dacBeepCodeTaskHandle = PIOS_Thread_Create(dacBeepCodeTask, "DacBeepCodes", STACK_SIZE_BYTES, NULL, TASK_PRIORITY);
+		TaskMonitorAdd(TASKINFO_RUNNING_ALTITUDEHOLD, dacBeepCodeTaskHandle); // TODO: delete, just stealing handle for debugging
```
TODO?

peabody124 replied:
Oh haha. I fixed that and didn't remove the TODO. Will update.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+		ManualControlCommandRssiGet(&rssi);
+		if (rssi < 0) rssi = 0;
+		if (rssi > 100) rssi = 100;
+
```
Extra newline?

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+
+		int16_t rssi = 0;
+		ManualControlCommandRssiGet(&rssi);
+		if (rssi < 0) rssi = 0;
```
Why this approach instead of ternary operator? Not that it matters, but I think in the past I've seen you use several ternary operators in cascaded succession so I'm wondering if and why you've changed preferences.

peabody124 replied:
Oh you are implying a lot more thought in my actions than is appropriate :) I bet you can find times where I've argued for one of them before though :)

peabody124 replied:
went to change this, but applying two conditionals on one line is ugly to me:
`rssi = rssi < 0 ? 0 : (rssi > 100 ? 100 : rssi)`
so just aesthetics.

kubark42 replied:
Fine by me.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+		PIOS_DACBEEP_Beep(dacbeep_handle, freq, 80);
+		PIOS_Thread_Sleep(period_ms);
+	}
+}
```
GitHub wants a newline.

kubark42 commented on flight/PiOS/STM32F4xx/inc/pios_dac_common.h:
```diff
@@ -0,0 +1,36 @@
+
+int32_t PIOS_DAC_COMMON_Init(void (*irq_cb_method)(void));
+
+#endif /* PIOS_DAC_COMMON_H */
```
GitHub wants a newline.

peabody124 replied:
done

kubark42 commented on flight/PiOS/STM32F4xx/pios_dacbeep.c:
```diff
@@ -0,0 +1,211 @@
+};
+
+struct pios_dacbeep_dev {
+	enum pios_dacbeep_dev_magic     magic;
```
Extra spaces before `magic`?

kubark42 commented on flight/PiOS/STM32F4xx/pios_dacbeep.c:
```diff
@@ -0,0 +1,211 @@
+#include "pios_dac_common.h"
+
+enum pios_dacbeep_dev_magic {
+	PIOS_DACBEEP_DEV_MAGIC = 0x3A53834A,
```
My favorite way to make magic device numbers is to run an md5 on the filename and then grab the first 8 digits. This is (almost) guaranteed to prevent collisions.

I find that making a note of it, like `PIOS_DACBEEP_DEV_MAGIC = 0x7e23b8ab // Magic number generated by md5 of piosdac_beep.c`, keeps my colleagues from guessing where it came from and lets them know how they can generate their own.

This isn't a blocker here, but if you like that approach feel free to use it.

peabody124 replied:
cool idea, might do that in the future. better than my manual attempt to generate entropy

kubark42 commented on flight/PiOS/STM32F4xx/pios_dacbeep.c:
```diff
@@ -0,0 +1,211 @@
+
+// Lookup tables for symbols
+// from ML designed to generate 0.4V p-p centered at 0.5V
+//   sprintf('%d,', round((sin(linspace(0,2*pi,130))*(2^12/3.3*0.2)+(2^12/3.3*0.5))))
```
Nice comment!

peabody124 replied:
Yeah - you should see the blocks of ML code embeded in the FSK_DAC code to come :)

kubark42 commented on flight/PiOS/inc/pios_dacbeep_priv.h:
```diff
@@ -0,0 +1,42 @@
+ * @brief PIOS interface for DAC Beep implementation
+ * @{
+ *
+ * @file       pios_dacbeep.h
```
`pios_dacbeep_priv.h`

kubark42 commented on flight/targets/sparky2/fw/pios_board.c:
```diff
@@ -683,13 +686,38 @@ void PIOS_Board_Init(void) {
+		PIOS_DACBEEP_Init(&dacbeep_id);
+		dacbeep_handle = dacbeep_id;
+	}
+#endif /* PIOS_INCLUDE_DAC_BEEPS */
```
Shouldn't there be an `#else` here in case someone chooses `HWSPARKY2_ADCDAC_BEEP` but `PIOS_INCLUDE_DAC_BEEPS` is _not_ included?

peabody124 replied:
What would you want the else to do? I could panic I suppose ...

kubark42 replied:
Yeah, I think that's appropriate. It's our normal way of saying "you configured something that can't be supported by the hardware". 

This is less important here than in situations where you might have support for multiple sensors but not all of them are built into the binary. We've got a board with 3 IMUs, 3 barometers, and 4 magnetometers. :)

kubark42 commented on flight/targets/sparky2/fw/pios_board.c:
```diff
@@ -683,13 +686,38 @@ void PIOS_Board_Init(void) {
+		uintptr_t fskdac_com_id;
+		PIOS_FSKDAC_Init(&fskdac_com_id, &pios_fskdac_config);
+	}
+#endif /* PIOS_INCLUDE_FSK */
```
Same question as above about the `#else`.

peabody124 replied:
this one I'll skip because FSK_DAC isn't actually included in this build - it's avoiding me bumping the UAVO twice.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+		// We want to map RSSI inversely into pulse duration. At 100% shoot for
+		// a period of 800 ms. At 50% get to 150ms. Sharp inflection at 50%.
+		// Using piecewise linear to avoid exponential calculation
+		int32_t period_ms = 100 + ((rssi < 50) ? rssi : (13 * (rssi - 50) + 50));
```
Should these values be configurable? Or if not that, then perhaps be `#define` instead of magic numbers?

peabody124 replied:
Let's go with define now. I think in the future people might want a lot more flexibility on how tones are generated but I'd rather defer that. For example we might want to do chirps because as @mlyle once pointed out people don't have absolute pitch.

kubark42 replied:
Configurability is cool, but I don't know about the absolute pitch problem. I suppose it depends on whether you're trying to communicate that something _is_, versus that something _is trending_.

Tying a variometer to an speaker is pretty standard with gliders, it does something similar to what you're doing here. If there's a problem with people identifying relative pitch they'll probably have discovered that already and published in the literature.

peabody124 replied:
Yeah and I mimicked this off a pulse ox when the important information is in pitch and that is encoded absolutely, so I'm ok for now. I just mean there are a whole variety of things people might want at some point to pump into an audio channel and I want to mainly get an infrastructure out for people.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+
+		// We want to map RSSI inversely into pulse duration. At 100% shoot for
+		// a period of 800 ms. At 50% get to 150ms. Sharp inflection at 50%.
+		// Using piecewise linear to avoid exponential calculation
```
You might add into the comment that `At 0%, get to 100ms`.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+		// Using piecewise linear to avoid exponential calculation
+		int32_t period_ms = 100 + ((rssi < 50) ? rssi : (13 * (rssi - 50) + 50));
+
+		// TODO: make battery scaling adjustable
```
Do you intend for this TODO to be completed before merge, or would you rather push off till some near future?

peabody124 replied:
Oh balls. That's very specific to BrushedSparky and should actually be fixed.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,119 @@
+		const float MAX_VOLT = 4.0f;
+		if (volt < MIN_VOLT) volt = MIN_VOLT;
+		if (volt > MAX_VOLT) volt = MAX_VOLT;
+		uint16_t freq = 1500 + (volt - MIN_VOLT) / (MAX_VOLT - MIN_VOLT) * 4000;
```
Need to handle case where `MIN_VOLT == MAX_VOLT`. Should also handle `MIN_VOLT > MAX_VOLT`, but that's less critical as it won't lead to a `NaN`.

kubark42 replied:
Should the frequency range be configurable as well?

kubark42 commented:
Neat stuff. Made several comments, but nothing major. Look forward to merging this.

peabody124 commented:
P.S. I'm going to mark this WIP and let's defer merging this. On the FSKDAC I can cause the board to lock up. Until I know it isn't in a shared part of this code, I don't want this going into next!

peabody124 committed:
DAC BEEP: whitespace/comment fixes from review

peabody124 commented:
Removed WIP. I confirmed why FSK DAC was having issues and it isn't because of this. I'll make those options configurable and this should be good for merging.

kubark42 commented:
@peabody124 were you done with this?

peabody124 commented:
@kubark42 sorry, was distracted this weekend on bug for tauosd but will do this shortly

kubark42 commented:
No prob, just wanted to be sure I wasn't merging something you were still working on. :)

peabody124 committed:
DAC BEEP: range pitch from alarm to warning voltage

This uses the alarm and warning thresholds in the
FlightBatterySettings object to determine the range
of audio pitches to use.

peabody124 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,129 @@
+		if (volt > MAX_VOLT) volt = MAX_VOLT;
+		uint16_t freq = MIN_FREQ_HZ;
+		if (MAX_VOLT > MIN_VOLT)
+			freq = MIN_FREQ_HZ + (volt - MIN_VOLT) / (MAX_VOLT - MIN_VOLT) * RANGE_FREQ_HZ;
```
@kubark42 updated and it is programmable now. I kept the frequency endpoints hardcoded for now and will wait for user feedback before many other changes.

kubark42 commented on flight/Modules/DacBeepCodes/dacbeepcodes.c:
```diff
@@ -0,0 +1,129 @@
+		const float MIN_VOLT = volt_threshold[FLIGHTBATTERYSETTINGS_VOLTAGETHRESHOLDS_ALARM];
+		const float MAX_VOLT = volt_threshold[FLIGHTBATTERYSETTINGS_VOLTAGETHRESHOLDS_WARNING];
+		const uint16_t MIN_FREQ_HZ = 1500;
+		const uint16_t RANGE_FREQ_HZ = 4000;
```
I'll merge as-is, but do you think it makes more sense to set a max and min frequency, instead of a min frequency and then range? Or, maybe even makes sense to define this in octaves, since most people won't realize that 1500Hz-5500Hz is much more compact than 150Hz-4150Hz?

peabody124 replied:
I started writing it as MAX but wanted to avoid the length of subtraction or an additional line, but will change it.

peabody124 committed:
DAC BEEP: specify start and stop frequencies

Using two octaves range to communicate.

peabody124 commented:
@kubark42 updated frequency boundaries, should be good to go

kubark42 commented:
@peabody124 I'm thinking about using this for beeps _a la_ PixHawk. Any thoughts on this? Should be as easy as connecting it to a speaker, although it might need some driver circuitry to get it loud enough to be heard.

peabody124 commented:
yeah also make sure you de-bias it because it hovers around 0.5V but most audio hardware should be AC i believe

kubark42 commented:
Good point. A simple capacitor in series should do the trick.
On Feb 18, 2016 8:51 PM, "peabody124" notifications@github.com wrote:

> yeah also make sure you de-bias it because it hovers around 0.5V but most
> audio hardware should be AC i believe
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/TauLabs/TauLabs/pull/2173#issuecomment-186008831.

kubark42 commented on flight/PiOS/STM32F4xx/pios_dac_common.c:
```diff
@@ -0,0 +1,112 @@
+	}
+
+#if defined(PIOS_INCLUDE_FREERTOS)
+	portEND_SWITCHING_ISR((rx_need_yield || tx_need_yield) ? pdTRUE : pdFALSE);
```
When compiling with FreeRTOS, this fails with unknown variables.

kubark42 commented on flight/PiOS/STM32F4xx/pios_dac_common.c:
```diff
@@ -0,0 +1,112 @@
+
+#include "pios.h"
+#if defined(PIOS_INCLUDE_DAC_BEEPS) || defined(PIOS_INCLUDE_FSK)
+
```
This needs

```
#if defined(PIOS_INCLUDE_FREERTOS)
#include "FreeRTOS.h"
#endif  // defined(PIOS_INCLUDE_FREERTOS)
```

in order to compile.
